File: smudge/utils/lastfm.py
import json
import re

from smudge.database.users import get_user_data, register_lastfm
from smudge.utils.utils import http

from ..config import config


class LastFM:
    """This class is a Python interface to Last.fm API."""

    def __init__(self):
        self.api: str = "http://ws.audioscrobbler.com/2.0"
        self.api_key: str = config["LASTFM_API_KEY"]
        self.is_connected: bool = False

    async def register_lastfm(self, id: int, username: str):
        if not await self.check_user(username):
            return False

        await register_lastfm(id, username)
        return True

    @classmethod
    async def get_username(self, id: int):
        username = (await get_user_data(id))["lastfm_username"]
        if not username:
            return False
        return username

    async def check_user(self, username: str):
        if not username:
            return False

        r = await http.get(
            self.api
            + "?method=user.getrecenttracks&limit=3"
            + f"&extended=1&user={username}"
            + f"&api_key={self.api_key}&format=json"
        )

        return r.status_code == 200

    async def track_playcount(self, artist: str, track: str, username: str):
        r = await http.get(
            self.api
            + f"?method=track.getinfo&artist={artist}"
            + f"&track={track}"
            + f"&user={username}&api_key={self.api_key}&format=json"
        )
        res = json.loads(r.content)

        try:
            scrobbles = int(res["track"]["userplaycount"]) + 1
        except KeyError:
            scrobbles = 1
        return scrobbles

    async def track(self, id: int):
        if not await self.check_user(await self.get_username(id)):
            return "No Username"

        r = await http.get(
            self.api
            + "?method=user.getrecenttracks&limit=3"
            + f"&extended=1&user={await self.get_username(id)}&api_key="
            + f"{self.api_key}&format=json"
        )

        if r.status_code != 200:
            return False

        res = json.loads(r.content)
        try:
            ftrack = res["recenttracks"]["track"][0]
        except IndexError:
            return "No Scrobbles"

        playcount = await self.track_playcount(
            ftrack["artist"]["name"], ftrack["name"], await self.get_username(id)
        )

        return {
            "artist": ftrack["artist"]["name"],
            "track": ftrack["name"],
            "loved": int(ftrack["loved"]),
            "playcount": playcount,
            "image": ftrack["image"][3]["#text"],
            "now": bool(ftrack.get("@attr")),
        }

    async def album_playcount(self, artist: str, album: str, username: str):
        r = await http.get(
            self.api
            + f"?method=album.getinfo&artist={artist}"
            + f"&album={album}"
            + f"&user={username}&api_key={self.api_key}&format=json"
        )
        res = json.loads(r.content)

        try:
            scrobbles = int(res["album"]["userplaycount"]) + 1
        except KeyError:
            scrobbles = 1
        return scrobbles

    async def album(self, id: int):
        if not await self.check_user(await self.get_username(id)):
            return "No Username"

        r = await http.get(
            self.api
            + "?method=user.getrecenttracks&limit=3"
            + f"&extended=1&user={await self.get_username(id)}&api_key="
            + f"{self.api_key}&format=json"
        )

        if r.status_code != 200:
            return False

        res = json.loads(r.content)
        try:
            ftrack = res["recenttracks"]["track"][0]
        except IndexError:
            return "No Scrobbles"

        playcount = await self.album_playcount(
            ftrack["artist"]["name"], ftrack["album"]["#text"], await self.get_username(id)
        )

        return {
            "artist": ftrack["artist"]["name"],
            "album": ftrack["album"]["#text"],
            "loved": int(ftrack["loved"]),
            "playcount": playcount,
            "image": ftrack["image"][3]["#text"],
            "now": bool(ftrack.get("@attr")),
        }

    async def artist_playcount(self, artist: str, username: str):
        r = await http.get(
            self.api
            + f"?method=artist.getInfo&artist={artist}"
            + f"&user={username}&api_key={self.api_key}&format=json"
        )
        res = json.loads(r.content)
        try:
            scrobbles = int(res["artist"]["stats"]["userplaycount"]) + 1
        except KeyError:
            scrobbles = 1
        return scrobbles

    async def artist(self, id: int):
        if not await self.check_user(await self.get_username(id)):
            return "No Username"

        r = await http.get(
            self.api
            + "?method=user.getrecenttracks&limit=3"
            + f"&extended=1&user={await self.get_username(id)}&api_key="
            + f"{self.api_key}&format=json"
        )

        if r.status_code != 200:
            return False

        res = json.loads(r.content)
        try:
            ftrack = res["recenttracks"]["track"][0]
        except IndexError:
            return "No Scrobbles"

        if found := re.search(
            'https://lastfm.freetls.fastly.net/i/u/avatar170s/.*?(?=")',
            (
                await http.get(
                    f"https://www.last.fm/music/{str(ftrack['artist']['name'])}/+images"
                )
            ).text,
        ):
            image = found.group().replace("avatar170s", "770x0") + ".jpg"
        else:
            image = ftrack["image"][3]["#text"]

        playcount = await self.artist_playcount(
            ftrack["artist"]["name"], await self.get_username(id)
        )
        return {
            "artist": ftrack["artist"]["name"],
            "loved": int(ftrack["loved"]),
            "playcount": playcount,
            "image": image,
            "now": bool(ftrack.get("@attr")),
        }

File: smudgelord/utils/lastfm/lastfm.go
package lastfm

import (
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"strconv"

	"smudgelord/smudgelord/config"
	"smudgelord/smudgelord/utils"
)

const lastFMAPI = "http://ws.audioscrobbler.com/2.0"

type LastFM struct {
	apiKey string
}

func Init() *LastFM {
	return &LastFM{apiKey: config.LastFMKey}
}

func (lfm *LastFM) GetUser(username string) error {
	body := utils.RequestGET(lastFMAPI, utils.RequestGETParams{Query: map[string]string{"method": "user.getinfo", "user": username, "api_key": lfm.apiKey, "format": "json"}})
	if body.StatusCode() != 200 {
		return nil
	}

	var userInfo userInfo
	err := json.Unmarshal(body.Body(), &userInfo)
	if err != nil {
		log.Print("[lastfm/GetUser] Error unmarshalling user info:", err)
	}

	if userInfo.User == nil {
		return errors.New("user not found")
	}
	return nil
}

func (lfm *LastFM) GetRecentTrack(username string) *recentTracks {
	body := utils.RequestGET(lastFMAPI, utils.RequestGETParams{Query: map[string]string{"method": "user.getrecenttracks", "user": username, "api_key": lfm.apiKey, "limit": "3", "extended": "1", "format": "json"}})
	if body.StatusCode() != 200 {
		return nil
	}

	var recentTracks recentTracks
	err := json.Unmarshal(body.Body(), &recentTracks)
	if err != nil {
		log.Print("[lastfm/GetRecentTrack] Error unmarshalling recent tracks:", err)
	}
	return &recentTracks
}

func (lfm *LastFM) PlayCount(recentTracks *recentTracks, method string) int {
	var methodValue string
	username := *recentTracks.RecentTracks.Attr.User // Dereference the pointer to get the string value
	artist := recentTracks.RecentTracks.Track[0].Artist.Name
	if method == "track" {
		methodValue = recentTracks.RecentTracks.Track[0].Name
	} else if method == "album" {
		methodValue = recentTracks.RecentTracks.Track[0].Album.Text
	} else if method == "artist" {
		methodValue = recentTracks.RecentTracks.Track[0].Artist.Name
	}

	body := utils.RequestGET(lastFMAPI, utils.RequestGETParams{Query: map[string]string{"method": fmt.Sprintf("%s.getinfo", method), "user": username, "api_key": lfm.apiKey, "artist": artist, method: methodValue, "format": "json"}}).Body()
	var getInfo getInfo
	err := json.Unmarshal(body, &getInfo)
	if err != nil {
		log.Print("[lastfm/PlayCount] Error unmarshalling get info:", err)
	}
	var userPlaycount int
	switch method {
	case "track":
		userPlaycount, err = strconv.Atoi(getInfo.Track.UserPlaycount)
	case "album":
		userPlaycount = getInfo.Album.Userplaycount
	case "artist":
		userPlaycount, err = strconv.Atoi(getInfo.Artist.Stats.Userplaycount)
	}
	if err != nil {
		return 0
	}

	if userPlaycount == 0 {
		return 1
	}

	return userPlaycount
}
